libphal/phal_sbe.C

```diff
@@ -370,5 +387,349 @@ void threadStopProc(struct pdbg_target *proc)
+		    seepromSide, pdbg_target_path(proc));
+		throw sbeError_t(exception::SBE_REIPL_WITHOUT_SWITCHING);
+	}
+
+	if (failCount >= MAX_SEEPROM_FAIL_COUNT) {
+		// RE-IPLing using the same side of the SEEPROM may not solve
+		// the issue, will attempt to switch the SEEPROM and re-ipl.
+		log(level::INFO, "RE-IPLing using the same side of the SEEPROM"
+				 " did not solve the issue. Will attempt to "
+				 "switch the SEEPROM"
+				 " and re-IPL");
+		(void)reiplUsingBKPBootSeeprom(proc);
+	}
+}
+
+/**
+ * @brief Switch the boot seeprom and REIPL using backup boot seeprom.
+ *
+ * @param[in] proc processor target to operate on
+ */
+void reiplUsingBKPBootSeeprom(struct pdbg_target *proc)
+{
+	assert(proc == NULL);
+
+	log(level::INFO, "reiplUsingBKPBootSeeprom on %s",
+	    pdbg_target_path(proc));
+
+	// Determine the current side of the SEEPROM used
+	uint8_t seepromSideUsed = getSeepromSideUsed(proc, SeepromType::Boot);
+
+	// update max failure count to the attribute to mark it as bad.
+	setSeepromFailCount(proc, SeepromType::Boot, seepromSideUsed,
+			    MAX_SEEPROM_FAIL_COUNT);
+
+	// Switch the seeprom if possible else throw error to log pel
+	// If current side of SEEPROM is primary , check the fail count of the
+	// secondary, vice-versa.
+	uint8_t failCount =
+	    getSeepromFailCount(proc, SeepromType::Boot, (!seepromSideUsed));
+	if (failCount < MAX_SEEPROM_FAIL_COUNT) {
+		setSeepromSideToUse(proc, SeepromType::Boot, !seepromSideUsed);
+		return;
+	}
+
+	log(level::ERROR, "Both primary and secondary boot seeproms are bad");
+	throw sbeError_t(exception::SBE_BOOT_SEEPROMS_BAD);
+}
+
+/**
+ * @brief Switch the measurement seeprom and REIPL using backup measurement
+ *        seeprom
+ * @param[in] proc processor target to operate on
+ */
+void reiplUsingBKPMeasurementSeeprom(struct pdbg_target *proc)
+{
+	log(level::INFO, "reiplUsingBKPMeasurementSeeprom on %s",
+	    pdbg_target_path(proc));
+	assert(proc == NULL);
+
+	log(level::INFO, "reiplUsingBKPMeasurementSeeprom on %s",
+	    pdbg_target_path(proc));
+
+	// Determine the current side of the SEEPROM used
+	uint8_t seepromSideUsed =
+	    getSeepromSideUsed(proc, SeepromType::Measurement);
+
+	// update max failure count to the attribute to mark it as bad.
+	setSeepromFailCount(proc, SeepromType::Measurement, seepromSideUsed,
+			    MAX_SEEPROM_FAIL_COUNT);
+
+	// Switch the seeprom if possible else throw error to log pel
+	// If current side of SEEPROM is primary , check the fail count of the
+	// secondary, vice-versa.
+	uint8_t failCount = getSeepromFailCount(proc, SeepromType::Measurement,
+						(!seepromSideUsed));
+	if (failCount < MAX_SEEPROM_FAIL_COUNT) {
+		setSeepromSideToUse(proc, SeepromType::Measurement,
+				    (!seepromSideUsed));
+		return;
+	}
+
+	log(level::ERROR,
+	    "Both primary and secondary measurement seeproms are bad");
+	throw sbeError_t(exception::SBE_MEASUREMENT_SEEPROMS_BAD);
+}
+
+void validatePrimaryProcTgt(struct pdbg_target *proc)
```

Comment:
Please add function headers for this and the following functions

Reply:
done